lua/orgmode/agenda/init.lua

```diff
@@ -457,16 +472,27 @@ function Agenda:agenda()
 
   for _, day in ipairs(dates) do
     local date = { day = day, agenda_items = {} }
+    local date_only = { day = day, agenda_items = {} }
+    local date_not_same = { day = day, agenda_items = {} }
 
     for _, item in ipairs(headline_dates) do
       local agenda_item = AgendaItem:new(item.headline_date, item.headline, day)
       if agenda_item.is_valid and self.filters:matches(item.headline) then
-        table.insert(date.agenda_items, agenda_item)
+        if not agenda_item.is_same_day then
+          table.insert(date_not_same.agenda_items, agenda_item)
+        elseif item.headline_date.date_only then
+          table.insert(date_only.agenda_items, agenda_item)
+        else
+          table.insert(date.agenda_items, agenda_item)
+        end
       end
     end
-
     date.agenda_items = sort_agenda_items(date.agenda_items)
+    date_not_same.agenda_items = sort_agenda_items(date_not_same.agenda_items)
+    date_only.agenda_items = sort_agenda_items_categories(date_only.agenda_items, category_inds)
 
+    date.agenda_items = utils.concat(date.agenda_items, date_only.agenda_items, false)
```

Comment:
@gerazov lets go with some solution to do all this sorting in the `sort_agenda_items` function. Saving all this additional state can potentially slow down agenda rendering, especially on month/year view.

Reply:
Ok I'll refactor :sweat_smile: